### ISSUE TITLE
Memory Consumption Higher with ZeRO GPT-2 vs other trainers

### ISSUE THREAD

**Issue by E-Bachman** [microsoft/DeepSpeed#485]
I am able to finetune [gpt-2-simple](https://github.com/minimaxir/gpt-2-simple) with model size of 774M using Google Colab. 

gpt-2-simple ``nvidia-smi: 15.4GB`` memory consumption on single T4 GPU

When using the [DeepSpeed/Megatron](https://github.com/microsoft/DeepSpeedExamples/tree/master/Megatron-LM) example the training consumes twice as much memory over multiple GPUs. 
 
> ZeRO Whitepaper "For a model such as GPT-2 with 1.5 Billion parameters, this leads toa memory requirement of at least 24GB"

For **774M** shouldn't it be comparable to ~6GB per GPU = 12GB. Am I missing something from the DeepSpeedExamples repo to allow *memory parallelism*?

```
774m
GPU1: 11GB/12Gb
GPU2: 11GB/12Gb
Total: 22GB/24Gb
```

```bash
NUM_WORKERS=1
NUM_GPUS_PER_WORKER=2
MP_SIZE=2

$ script \
       --model-parallel-size 2 \
       --num-layers 36 \
       --hidden-size 1280 \
       --num-attention-heads 20 \
       --batch-size 1 \
       --seq-length 1024 \
       --max-position-embeddings 1024 \
       --tokenizer-type GPT2BPETokenizer \
       --split 949,50,1 \
       --distributed-backend nccl \
       --lr 0.00015 \
       --no-load-optim \
       --lr-decay-style cosine \
       --weight-decay 1e-2 \
       --clip-grad 1.0 \
       --warmup .01 \
       --checkpoint-activations \
       --deepspeed-activation-checkpointing \
       --fp16 \
       --deepspeed \
       --deepspeed_config scripts/ds_zero2_config.json
```

Modified from [scripts/ds_zero2_config.json](https://github.com/microsoft/DeepSpeedExamples/blob/master/Megatron-LM/scripts/ds_zero2_config.json) with every bell and whistle for optimization.

```json
{
  "train_batch_size": 16,
  "gradient_accumulation_steps": 1,
  "steps_per_print": 1,
  "zero_optimization": {
    "stage": 2,
    "allgather_partitions": true,
    "contiguous_gradients": true,
    "overlap_comm": true,
    "reduce_scatter": true,  
    "reduce_bucket_size": 50000000,
    "allgather_bucket_size": 500000000
  },
  "zero_allow_untested_optimizer": true,
  "gradient_clipping": 1.0,
  "fp16": {
    "enabled": true,
    "loss_scale": 0,
    "loss_scale_window": 1000,
    "hysteresis": 2,
    "min_loss_scale": 1
  },
  "activation_checkpointing": {
    "partition_activations": true,
    "contiguous_memory_optimization": true
  },
  "wall_clock_breakdown": false
}
```

**Comment by tjruwase** [comment]
@E-Bachman I suspect that the bucket size values are responsible for the memory bloat. As suggested in #467, can you try reducing these to 1e7 from 5e8?

**Comment by E-Bachman** [comment]
> @E-Bachman I suspect that the bucket size values are responsible for the memory bloat. As suggested in #467, can you try reducing these to 1e7 from 5e8?

``1e7`` and ``overlap_comm: false`` allows 774M to fit nicely

```
nvidia-smi
GPU0: 8.7G/12G
GPU1: 8.7G/12G
Total: 17.4G/24G
```

@tjruwase Was there any other tweaks to get 1.5B model to run in 24G? Getting CUDA OOM ``tried to allocate 2.6GB with 9.41GB reserved and 1.96GB Free``

**Comment by tjruwase** [comment]
@E-Bachman, we did not use ZeRO-stage1 to train a 1.5B model on 24GB. What that statement in the paper is saying is that for a 1.5B model, 24GB is required to store fp16 weights/gradients (6GB) and Adam optimizer state (18GB). This is because Adam stores fp32 master weight, momentum, and variance for each parameter: 3 * 4Bytes * 1.5Billion = 18GB. All the experiments in the paper are on 32GB V100s.  

If you need to run larger models you can try out [ZeRO-Offload](https://www.deepspeed.ai/tutorials/zero-offload/) or [ZeRO-stage2](https://www.deepspeed.ai/tutorials/zero/#training-a-10b-parameter-gpt-2-model).

**Comment by tjruwase** [comment]
Closing for lack of activity. Please reopen as needed.
